AX: Store AXIsolatedObject::{m_platformWidget, m_remoteParent} as properties to avoid paying the memory cost for them on every single object

twilco opened the pull request:
#### 0378dfedf7b50590fa8309aef54da27002e6bb9f
<pre>
AX: Store AXIsolatedObject::{m_platformWidget, m_remoteParent} as properties to avoid paying the memory cost for them on every single object
<a href="https://bugs.webkit.org/show_bug.cgi?id=292689">https://bugs.webkit.org/show_bug.cgi?id=292689</a>
<a href="https://rdar.apple.com/150887081">rdar://150887081</a>

Reviewed by Joshua Hoffman.

Only scroll views and attachments ever set AXIsolatedObject::{m_platformWidget, m_remoteParent}, but because they are
member variables, we pay the cost for them on every single object. We can have hundreds of thousands of instances of
this class, so that&apos;s a major cost.

This commit turns these fields into AXProperty::PlatformWidget and AXProperty::RemoteParent. This commit also rearranges
AXIsolatedObject&apos;s fields to be more compact, avoiding unnecessary padding.

Prior to this commit, AXIsolatedObject was 136 bytes. Now it&apos;s 112 bytes. This saves 13.84mb on <a href="https://html.spec.whatwg.org">https://html.spec.whatwg.org</a>,
which has 567769 objects.

* Source/WebCore/accessibility/AXCoreObject.h:
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::remoteParentObject const): Deleted.
* Source/WebCore/accessibility/AccessibilityObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::setProperty):
(WebCore::AXIsolatedObject::platformWidget const):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
* Source/WebCore/accessibility/isolatedtree/mac/AXIsolatedObjectMac.mm:
(WebCore::AXIsolatedObject::initializePlatformProperties):
(WebCore::AXIsolatedObject::remoteParent const):
(WebCore::AXIsolatedObject::remoteParentObject const): Deleted.
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(-[WebAccessibilityObjectWrapper remoteAccessibilityParentObject]):
(scrollViewParent):

Canonical link: <a href="https://commits.webkit.org/294666@main">https://commits.webkit.org/294666@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a807f6d51ad08974f1f01857a8b3bbcd50118e3d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/102582 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/22251 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/12565 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/107743 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/53219 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/104621 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/22559 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/30757 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/78033 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/35011 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/105588 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/17477 "Passed tests") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/92588 "Found 1 new API test failure: TestWebKitAPI.WebKit.PrintFrame (failure)") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/58368 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/17316 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/10617 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/52576 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/87126 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/10686 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/110118 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/29714 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/21910 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/87010 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/30078 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/88783 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/86616 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22053 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/31439 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/9169 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/23964 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/29642 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/34954 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/29453 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/32776 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/31012 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

minorninth commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
```diff
@@ -571,21 +571,20 @@ class AXIsolatedObject final : public AXCoreObject {
-    RetainPtr<RemoteAXObjectRef> m_remoteParent;
-#else
+#if !PLATFORM(COCOA)
     PlatformWidget m_platformWidget;
```
Is this a candidate for a similar future optimization?

minorninth replied:
Oh, nevermind - this would be for platforms like GTK.

minorninth reviewed (approved):
lgtm!

twilco committed:
AX: Store AXIsolatedObject::{m_platformWidget, m_remoteParent} as properties to avoid paying the memory cost for them on every single object
https://bugs.webkit.org/show_bug.cgi?id=292689
rdar://150887081

Reviewed by Joshua Hoffman.

Only scroll views and attachments ever set AXIsolatedObject::{m_platformWidget, m_remoteParent}, but because they are
member variables, we pay the cost for them on every single object. We can have hundreds of thousands of instances of
this class, so that's a major cost.

This commit turns these fields into AXProperty::PlatformWidget and AXProperty::RemoteParent. This commit also rearranges
AXIsolatedObject's fields to be more compact, avoiding unnecessary padding.

Prior to this commit, AXIsolatedObject was 136 bytes. Now it's 112 bytes. This saves 13.84mb on https://html.spec.whatwg.org,
which has 567769 objects.

* Source/WebCore/accessibility/AXCoreObject.h:
* Source/WebCore/accessibility/AXLogger.cpp:
(WebCore::operator<<):
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::remoteParentObject const): Deleted.
* Source/WebCore/accessibility/AccessibilityObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::setProperty):
(WebCore::AXIsolatedObject::platformWidget const):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:
* Source/WebCore/accessibility/isolatedtree/mac/AXIsolatedObjectMac.mm:
(WebCore::AXIsolatedObject::initializePlatformProperties):
(WebCore::AXIsolatedObject::remoteParent const):
(WebCore::AXIsolatedObject::remoteParentObject const): Deleted.
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(-[WebAccessibilityObjectWrapper remoteAccessibilityParentObject]):
(scrollViewParent):

Canonical link: https://commits.webkit.org/294666@main
